servant-server/src/Servant/Server/Internal/RoutingApplication.hs

```diff
@@ -155,32 +159,111 @@ data Delayed :: * -> * where
   Delayed :: IO (RouteResult a)
           -> IO (RouteResult ())
           -> IO (RouteResult b)
-          -> (a -> b -> RouteResult c)
-          -> Delayed c
+          -> IO (RouteResult c)
+          -> (a -> b -> c -> RouteResult d)
+          -> Delayed d
 
 instance Functor Delayed where
-   fmap f (Delayed a b c g) = Delayed a b c ((fmap.fmap.fmap) f g)
+   fmap f (Delayed a b c d g) = Delayed a b c d ((fmap.fmap.fmap.fmap) f g)
 
 -- | Add a capture to the end of the capture block.
 addCapture :: Delayed (a -> b)
            -> IO (RouteResult a)
            -> Delayed b
-addCapture (Delayed captures method body server) new =
-  Delayed (combineRouteResults (,) captures new) method body (\ (x, v) y -> ($ v) <$> server x y)
+addCapture (Delayed captures method auth body server) new =
+  Delayed (combineRouteResults (,) captures new) method auth body (\ (x, v) y z -> ($ v) <$> server x y z)
 
 -- | Add a method check to the end of the method block.
 addMethodCheck :: Delayed a
                -> IO (RouteResult ())
                -> Delayed a
-addMethodCheck (Delayed captures method body server) new =
-  Delayed captures (combineRouteResults const method new) body server
+addMethodCheck (Delayed captures method auth body server) new =
+  Delayed captures (combineRouteResults const method new) auth body server
+
+-- | helper type family to capture server handled values for various policies
+type family AuthDelayedReturn (mP :: AuthPolicy) mE (uP :: AuthPolicy) uE usr :: * where
+    AuthDelayedReturn 'Strict mE 'Strict uE usr = usr
+    AuthDelayedReturn 'Strict mE 'Lax    uE usr = Either uE usr
+    AuthDelayedReturn 'Lax    mE 'Strict uE usr = Either mE usr
+    AuthDelayedReturn 'Lax    mE 'Lax    uE usr = Either (Either mE uE) usr
+
+-- | Internal method to generate auth checkers for various policies. Scary type signature
+-- but it does help with understanding the logic of how each policy works. See
+-- examples below.
+genAuthCheck :: (OnMissing IO ServantErr mPolicy mError -> (AuthDelayedReturn mPolicy mError uPolicy uError usr -> a) -> mError -> IO (RouteResult a))
+             -> (OnUnauthenticated IO ServantErr uPolicy uError auth -> (AuthDelayedReturn mPolicy mError uPolicy uError usr -> a) -> uError -> auth -> IO (RouteResult a))
+             -> (usr -> (AuthDelayedReturn mPolicy mError uPolicy uError usr))
+             -> Delayed (AuthProtected IO ServantErr mPolicy mError uPolicy uError auth usr (AuthDelayedReturn mPolicy mError uPolicy uError usr -> a))
+             -> IO (RouteResult (Either mError auth))
+             -> Delayed a
+genAuthCheck missingHandler unauthHandler returnHandler d@(Delayed captures method _ body _) new =
+    let newAuth =
+            runDelayed d `bindRouteResults` \ authProtection ->
```

Comment:
So, to be honest, my understanding of how `Delayed` is used is still a bit shaky. For example, it seems like anyone using `runDelayed` must provide a function that accepts authentication params (which could be `()` but I don't see that being used anywhere).

To your question: this can change, but as it stands I've placed the order as: `captures`, `method`, `auth`, `body`. So, `captures` and `method` will run first, then `auth`, then `body`.

As for destroying the order setup, my comment was this: if we add an `authCheck` what we are returning is a `Delayed` with an authentication param in it. So, when we eventually `runDelayed` the order is restored, because we run first the `captures`, then the `method`, then the `auth`, and finally the `body` checks.

Now, if I understand things correctly, what I think happens because we need to extract the `AuthProtected` data is that the `captures` and `body` checks happen twice, which is unfortunate.